# Skyline Enrollments: Hypothesis Tests

Lesson 1.6 Practice. Runs t-tests on the Skyline Enrollments dataset,
demonstrates the multiple testing problem with simulation, and shows how Bonferroni correction works.

Author: [A. Chris Yi]
Date: [2026-09-23]

In [71]:
import numpy as np
import pandas as pd
from scipy import stats

In [72]:
enrollments = pd.read_csv("../lesson-1-2-types-of-data/skyline_enrollments.csv")

python_for_beginners = enrollments[enrollments["course_name"] == "Python for Beginners"]["hours_studied"]
sql_basics = enrollments[enrollments["course_name"] == "SQL Basics"]["hours_studied"]

print(f"Python for Beginners: n={len(python_for_beginners)}, mean={python_for_beginners.mean():.2f}, std={python_for_beginners.std():.2f}")
print(f"SQL Basics: n={len(sql_basics)}, mean={sql_basics.mean():.2f}, std={sql_basics.std():.2f}")


Python for Beginners: n=23, mean=32.46, std=16.85
SQL Basics: n=17, mean=35.80, std=12.76


In [73]:
t_stat, p_value = stats.ttest_ind(python_for_beginners, sql_basics)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: -0.6845
p-value: 0.4978


In [74]:
mean_diff = python_for_beginners.mean() - sql_basics.mean()

n1, n2 = len(python_for_beginners), len(sql_basics)
s1, s2 = python_for_beginners.std(ddof=1), sql_basics.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (Python for Beginners - SQL Basics): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")
print(f"p-value: {p_value:.4f}")

Mean difference (Python for Beginners - SQL Basics): -3.34
95% CI for the difference: (-12.52, 5.84)
p-value: 0.4978


The mean hours studied for SQL Basics exceeded that of Python for Beginners by 3.34 hours (95% CI: -12.52 to 5.84), p > 0.49. While zero falls in the CI, the p-value is near 0.5, so the difference is not statistically significant and in reality it's a comprable chance to flipping a coin. Decisions should not be made on this data alone. More data points and further analysis are needed to determine whether there is an identifiable difference in the courses in terms of study hours needed.

In [75]:
statistics_101 = enrollments[enrollments["course_name"] == "Statistics 101"]["hours_studied"]
tableau_fundamentals = enrollments[enrollments["course_name"] == "Tableau Fundamentals"]["hours_studied"]
intro_to_analytics = enrollments[enrollments["course_name"] == "Intro to Analytics"]["hours_studied"]

print(f"Statistics 101: n={len(statistics_101)}, mean={statistics_101.mean():.2f}, std={statistics_101.std():.2f}")
print(f"Tableau Fundamentals: n={len(tableau_fundamentals)}, mean={tableau_fundamentals.mean():.2f}, std={tableau_fundamentals.std():.2f}")
print(f"Intro to Analytics: n={len(intro_to_analytics)}, mean={intro_to_analytics.mean():.2f}, std={intro_to_analytics.std():.2f}")



Statistics 101: n=18, mean=26.03, std=14.97
Tableau Fundamentals: n=26, mean=30.41, std=17.01
Intro to Analytics: n=16, mean=28.06, std=17.15


In [76]:
t_stat, p_value = stats.ttest_ind(python_for_beginners, statistics_101)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 1.2728
p-value: 0.2106


In [77]:
t_stat, p_value = stats.ttest_ind(python_for_beginners, tableau_fundamentals)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 0.4235
p-value: 0.6738


In [78]:
t_stat, p_value = stats.ttest_ind(python_for_beginners, intro_to_analytics)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 0.7970
p-value: 0.4305


In [79]:
t_stat, p_value = stats.ttest_ind(sql_basics, statistics_101)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 2.0728
p-value: 0.0461


In [80]:
t_stat, p_value = stats.ttest_ind(sql_basics, tableau_fundamentals)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 1.1164
p-value: 0.2707


In [81]:
t_stat, p_value = stats.ttest_ind(sql_basics, intro_to_analytics)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 1.4779
p-value: 0.1495


In [82]:
t_stat, p_value = stats.ttest_ind(statistics_101, tableau_fundamentals)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: -0.8805
p-value: 0.3836


In [83]:
t_stat, p_value = stats.ttest_ind(statistics_101, intro_to_analytics)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: -0.3681
p-value: 0.7152


In [84]:
t_stat, p_value = stats.ttest_ind(tableau_fundamentals, intro_to_analytics)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 0.4335
p-value: 0.6670


In [85]:
mean_diff = python_for_beginners.mean() - statistics_101.mean()

n1, n2 = len(python_for_beginners), len(statistics_101)
s1, s2 = python_for_beginners.std(ddof=1), statistics_101.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (Python for Beginners - Statistics 101): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (Python for Beginners - Statistics 101): 6.43
95% CI for the difference: (-3.33, 16.19)


In [86]:
mean_diff = python_for_beginners.mean() - tableau_fundamentals.mean()

n1, n2 = len(python_for_beginners), len(tableau_fundamentals)
s1, s2 = python_for_beginners.std(ddof=1), tableau_fundamentals.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (Python for Beginners - Tableau Fundamentals): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (Python for Beginners - Tableau Fundamentals): 2.05
95% CI for the difference: (-7.44, 11.55)


In [87]:
mean_diff = python_for_beginners.mean() - intro_to_analytics.mean()

n1, n2 = len(python_for_beginners), len(intro_to_analytics)
s1, s2 = python_for_beginners.std(ddof=1), intro_to_analytics.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (Python for Beginners - Intro to Analytics): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (Python for Beginners - Intro to Analytics): 4.40
95% CI for the difference: (-6.46, 15.27)


In [88]:
mean_diff = sql_basics.mean() - statistics_101.mean()

n1, n2 = len(sql_basics), len(statistics_101)
s1, s2 = sql_basics.std(ddof=1), statistics_101.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (SQL Basics - Statistics 101): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (SQL Basics - Statistics 101): 9.77
95% CI for the difference: (0.57, 18.97)


In [89]:
mean_diff = sql_basics.mean() - tableau_fundamentals.mean()

n1, n2 = len(sql_basics), len(tableau_fundamentals)
s1, s2 = sql_basics.std(ddof=1), tableau_fundamentals.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (SQL Basics - Tableau Fundamentals): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (SQL Basics - Tableau Fundamentals): 5.39
95% CI for the difference: (-3.53, 14.31)


In [90]:
mean_diff = sql_basics.mean() - intro_to_analytics.mean()

n1, n2 = len(sql_basics), len(intro_to_analytics)
s1, s2 = sql_basics.std(ddof=1), intro_to_analytics.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (SQL Basics - Intro to Analytics): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (SQL Basics - Intro to Analytics): 7.75
95% CI for the difference: (-2.62, 18.11)


In [91]:
mean_diff = statistics_101.mean() - tableau_fundamentals.mean()

n1, n2 = len(statistics_101), len(tableau_fundamentals)
s1, s2 = statistics_101.std(ddof=1), tableau_fundamentals.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (Statistics 101 - Tableau Fundamentals): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (Statistics 101 - Tableau Fundamentals): -4.38
95% CI for the difference: (-13.89, 5.14)


In [92]:
mean_diff = statistics_101.mean() - intro_to_analytics.mean()

n1, n2 = len(statistics_101), len(intro_to_analytics)
s1, s2 = statistics_101.std(ddof=1), intro_to_analytics.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (Statistics 101 - Intro to Analytics): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (Statistics 101 - Intro to Analytics): -2.03
95% CI for the difference: (-12.91, 8.85)


In [93]:
mean_diff = tableau_fundamentals.mean() - intro_to_analytics.mean()

n1, n2 = len(tableau_fundamentals), len(intro_to_analytics)
s1, s2 = tableau_fundamentals.std(ddof=1), intro_to_analytics.std(ddof=1)
se_diff = np.sqrt(s1**2 / n1 + s2**2 / n2)

# 95% CI for the difference
margin = 1.96 * se_diff
ci_lower = mean_diff - margin
ci_upper = mean_diff + margin

print(f"Mean difference (Tableau Fundamentals - Intro to Analytics): {mean_diff:.2f}")
print(f"95% CI for the difference: ({ci_lower:.2f}, {ci_upper:.2f})")


Mean difference (Tableau Fundamentals - Intro to Analytics): 2.35
95% CI for the difference: (-8.30, 13.00)


In [95]:
import pandas as pd
df = pd.read_csv("lesson-1-6-hypothesis-testing-results.csv")
display(df)

,course_a,course_b,mean_difference,p_value,Unnamed: 4
0,Python for Beginners,SQL Basics,-3.34,0.4978,NaN
1,Python for Beginners,Statistics 101,6.43,0.2106,NaN
2,Python for Beginners,Tableau Foundations,2.05,0.4235,NaN
3,Python for Beginners,Intro to Analytics,4.40,0.4305,NaN
4,SQL Basics,Statistics 101,9.77,0.0461,NaN
5,SQL Basics,Tableau Foundations,5.39,0.2707,NaN
6,SQL Basics,Intro to Analytics,7.75,0.1495,NaN
7,Statistics 101,Tableau Foundations,-4.38,0.3836,NaN
8,Statistics 101,Intro to Analytics,-2.03,0.7152,NaN
9,Tableau Foundations,Intro to Analytics,2.35,0.6670,NaN


Only 1 test (SQL Basics and Statistics 101) came back with a p < 0.05. 

The Bonferroni correction for the 10 tests would be 0.005. 

No tests survived the Bonferroni correction. 

Given that the data was generated with similar distributions, I should see no significant results 95% of the time. The one test that came back as significant before Bonferroni correction was proven to be data noise and not actually significant upon correction. The variation in amount of hours spent studying is purely chance. 